pyomo/repn/plugins/parameterized_standard_form.py

```diff
@@ -0,0 +1,233 @@
+            default=None,
+            domain=var_component_set,
+            description="Vars to treat as data for the purposes of compiling"
+            "the standard form",
+            doc="""
+            Optional list of Vars to be treated as data while compiling the 
+            standard form.
+
+            For example, if this is the standard form of an inner problem in a
+            multilevel optimization problem, then the outer problem's Vars would
+            be specified in this list since they are not variables from the
+            perspective of the inner problem.
+            """,
+        ),
+    )
+
+    @document_kwargs_from_configdict(CONFIG)
+    def write(self, model, ostream=None, **options):
+        """Convert a model to standard form (`min cTx s.t. Ax <= b`) treating the
+        Vars specified in 'wrt' as data
+
+        Returns
+        -------
+        LinearStandardFormInfo
+
+        Parameters
+        ----------
+        model: ConcreteModel
+            The concrete Pyomo model to write out.
+
+        ostream: None
+            This is provided for API compatibility with other writers
+            and is ignored here.
+
+        """
+        config = self.config(options)
+
+        # Pause the GC, as the walker that generates the compiled LP
+        # representation generates (and disposes of) a large number of
+        # small objects.
+        with PauseGC():
+            return _ParameterizedLinearStandardFormCompiler_impl(config).write(model)
+
+
+class _SparseMatrixBase(object):
+    def __init__(self, matrix_data, shape):
+        (data, indices, indptr) = matrix_data
+        (nrows, ncols) = shape
+
+        self.data = np.array(data)
+        self.indices = np.array(indices, dtype=int)
+        self.indptr = np.array(indptr, dtype=int)
+        self.shape = (nrows, ncols)
+
+    def __eq__(self, other):
+        return self.todense() == other
+
+
+class _CSRMatrix(_SparseMatrixBase):
+    def tocsc(self):
+        # Implements the same algorithm as scipy's csr_tocsc function
+        csr_data = self.data
+        col_index = self.indices
+        row_index_ptr = self.indptr
+        nrows = self.shape[0]
+
+        num_nonzeros = len(csr_data)
+        csc_data = np.empty(csr_data.shape[0], dtype=object)
+        row_index = np.empty(num_nonzeros, dtype=int)
+        # tally the nonzeros in each column
+        col_index_ptr = np.zeros(self.shape[1], dtype=int)
+        for i in col_index:
+            col_index_ptr[int(i)] += 1
+
+        # cumulative sum the tally to get the column index pointer
+        cum_sum = 0
+        for i, tally in enumerate(col_index_ptr):
+            col_index_ptr[i] = cum_sum
+            cum_sum += tally
+        # we leave off the last entry (the number of nonzeros) because we are
+        # going to do the cute scipy thing and it will get 'added' at the end
+        # when we shift right (by which I mean it will conveniently already be
+        # there)
```

Comment:
It's more commenting what we're *not* doing at this point in the code--the `col_index_ptr` is missing the last element. And that gets "fixed" in the code block below and in line 139 below. This comment is really just documenting the algorithm because it's very cute but also very confusing to read.

Reply:
You could make this a little more clear by updating the comment to something like:
```
We have now initialized the col_index_ptr to the *starting* position of
each column in the data vector.  Note that col_index_ptr is only
num_cols long: we have ignored the last entry in the standard CSC
col_index_ptr (the total number of nonzeros).  This will get resolved
below when we shift this vector by one position.
```
And then below with something like
```
Now we are actually going to mess up what we just did while we construct
the row index: We can imagine that col_index_ptr holds the position of
the *next* nonzero in each column, so each time we move a data element
into a column we will increment that col_index_ptr by one.  This is
beautiful because by "messing up" the col_index_pointer, we are just
transforming the vector of *starting* indices for each column to a
vector of *ending* indices (actually 1 past the last index) of each
column.  Thank you, scipy.
```
(and maybe even update further below)
```
Fix the column index pointer by inserting 0 at the beginning.  The
col_index_ptr currently holds pointers to 1 past the last element of
each column, which is really the starting index for the next column.
Inserting the 0 (the starting index for the firsst column) shifts
everything by one column, "converting" the vector to the starting
indices of each column, and extending the vector length to num_cols + 1
(as is expected by the CSC matrix).
```